src/Integration.UnitTests/Helpers/MefFactoryTests.cs

```diff
@@ -0,0 +1,103 @@
+using Moq;
+using SonarLint.VisualStudio.Core;
+using SonarLint.VisualStudio.Integration.Helpers;
+using SonarLint.VisualStudio.TestInfrastructure;
+
+namespace SonarLint.VisualStudio.Integration.UnitTests.Helpers
+{
+    [TestClass]
+    public class MefFactoryTests
+    {
+        [TestMethod]
+        public void MefCtor_CheckIsExported()
+        {
+            MefTestHelpers.CheckTypeCanBeImported<MefFactory, IMefFactory>(
+                MefTestHelpers.CreateExport<SVsServiceProvider>(),
+                MefTestHelpers.CreateExport<IThreadHandling>());
+        }
+
+        [TestMethod]
+        public void CheckIsSingletonMefComponent()
+        {
+            MefTestHelpers.CheckIsSingletonMefComponent<MefFactory>();
+        }
+
+        [TestMethod]
+        public async Task CreateAsync_IsCalledOnMainThread()
+        {
+            var serviceProvider = SetUpServiceProviderWithComponentModel();
+            var threadHandling = new Mock<IThreadHandling>();
+            Action runOnUiAction = null;
+            threadHandling
+                .Setup(x => x.RunOnUIThread(It.IsAny<Action>()))
+                .Callback((Action callbackAction) => runOnUiAction = callbackAction);
+
+            var testSubject = new MefFactory(serviceProvider.Object, threadHandling.Object);
+            await testSubject.CreateAsync<NonSharedEmptyTestMefImplementation>();
+
+            threadHandling.Verify(x => x.RunOnUIThread(It.IsAny<Action>()), Times.Once);
+
+            serviceProvider.Invocations.Should().HaveCount(0);
+
+            runOnUiAction.Should().NotBeNull();
+            runOnUiAction();
+
+            serviceProvider.Verify(x => x.GetService(typeof(SComponentModel)), Times.Once);
+        }
+
+        [TestMethod]
+        public async Task CreateAsync_ReturnCorrectObject()
+        {
+            var testObject = new NonSharedEmptyTestMefImplementation();
+
+            var serviceProvider = SetUpServiceProviderWithComponentModel(testObject);
+
+            var testSubject = new MefFactory(serviceProvider.Object, new NoOpThreadHandler());
+            var result = await testSubject.CreateAsync<NonSharedEmptyTestMefImplementation>();
+
+            result.Should().Be(testObject);
+        }
+
+        private Mock<IServiceProvider> SetUpServiceProviderWithComponentModel(NonSharedEmptyTestMefImplementation testObject = null)
+        {
+            var serviceProvider = new Mock<IServiceProvider>();
+            var componentModel = new Mock<IComponentModel>();
+
+            componentModel.Setup(x => x.GetService<NonSharedEmptyTestMefImplementation>()).Returns(testObject);
+            serviceProvider.Setup(x => x.GetService(typeof(SComponentModel))).Returns(componentModel.Object);
+
+            return serviceProvider;
+        }
+
+        [Export(typeof(NonSharedEmptyTestMefImplementation))]
+        [PartCreationPolicy(CreationPolicy.NonShared)]
+        internal class NonSharedEmptyTestMefImplementation { }
```

Comment:
I am using this to test the implementation of the CreateAsync so that we do not have to use a random mef class